### PR TITLE
Don't crash when reading emoji input in utf8

### LINKED ISSUE #8663: OpenConsole crash when pasting emoji
# Environment

```none
Windows build number:  10.0.19041.630]
Windows Terminal version (if applicable):

Any other software?
tcsh (native Windows version)

# Steps to reproduce

download tcsh-x64.exe from https://github.com/amoldeshpande/tcsh/releases/tag/20210103.10 

Launch tcsh.  Doesn't have to be in terminal.  Paste any emoji into it.  It will crash conhost.exe on Windows 10 2004 and OpenConsole.exe if you launch within Windows Terminal

# Expected behavior

Should paste an emoji

# Actual behavior


Edited: repro steps with tcsh. Cannnot repro it with my test program anymore

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

In the two places we read input from the console input buffer, we were not accounting for the fact that unicode input might turn into more than two input records.

`InputBuffer::Read` tries to account for how big unicode characters will be when read as input events, and only returns as much input as _will_ fit in the caller's buffer. However, we were using `IsGlyphFullWidth` to check if a glyph was wide. However, that's not always right. `SplitToOem` might make three `char`s out of a single key, like in the case of emoji in utf8. 

## PR Checklist
* [x] Closes #8663
* [x] I work here
* [x] Tests added/passed
* [n/a] Requires documentation to be updated

## Validation Steps Performed
* wrote tests
* Confirmed that the sample in https://github.com/microsoft/terminal/issues/8663#issuecomment-904788661 no longer crashes

### CODE REVIEW

File: src/host/inputBuffer.cpp [hunk 1]
```diff
@@ -407,24 +410,45 @@ void InputBuffer::_ReadBuffer(_Out_ std::deque<std::unique_ptr<IInputEvent>>& ou
...
+
+                // convert from wchar to char
+                std::wstring wstr{ pKeyEvent->GetCharData() };
+                const auto str = ConvertToA(codepage, wstr);
+                sizeOfThisKey = str.size();
+                if (virtualReadCount + sizeOfThisKey > readCount)
                 {
-                    ++virtualReadCount;
+                    break;
                 }
```
Comment (lhecker) [inline src/host/inputBuffer.cpp:434]: This loop now has two bounds checks instead of one, which might be a bit confusing. We can simplify it by changing the loop to be `while (!_storage.empty())` and replacing this if condition with the following:
```cpp
if (!unicode)
{
    ...
}

virtualReadCount += sizeOfThisKey;
if (virtualReadCount > readCount)
{
    break;
}

if (performNormalRead)
{
    ...
}
```

**Review by lhecker — COMMENTED** [review]
I'd approve this code, but I don't get what `virtualReadCount` is supposed to be. The comment for it implies it counts the number of columns ("DBCS records". Right?), which matches the old code, but the new code counts code points. What if the CP is 65001?
Finally the comment for the `readCount` parameter implies it counts the number of events, but if that was true we'd only increment `virtualReadCount` whenever we push into the `outEvents` vector.

In other words: I don't really get the code. 😅

**Comment by miniksa** [comment]
The test is now stuck in an infinite loop so something's wrong here.

**Comment by carlos-zamora** [comment]
Circling back on this PR. Assigning @zadjii-msft since the test is stuck on an infinite loop.

**Comment by lhecker** [comment]
#14745 is an alternative approach. I've tested it with the repros in #8663 and it works great.

**Comment by DHowett** [comment]
@zadjii-msft, mind if we close this one in favor of Leonard's rewrite?

**Comment by zadjii-msft** [comment]
please yes
